### PR TITLE
Cartage 2.0

### PR DESCRIPTION
**Pull request by halostatue**
-   Substantial restructuring; mostly not backwards compatible with Cartage 1.0.
  -   Add compression configuration.
  -   Allow disabling dependency caching.
  -   Switch from release_hashref text file to release-metadata JSON file.
  -   New plug-in architecture for plug-ins that can insert themselves into the execution of code.
  -   Cleaned up some of the rough edges on Cartage::Config
    -   Remove any support for environment configs.
    -   Remove user-local and global Cartage configuration support. It was silly.
-   Switch from CmdParse to GLI. Old commands must be rewritten to the new format.
-   Start using more modern Ruby features and backporting features that can be simulated with pure Ruby.
-   Update documentation for Cartage 2.0.
  -   Document writing Cartage extensions in more depth in Extending.md.
  -   Mix Markdown and Rdoc-format documentation because Rdoc is not great for longer prose-style documentation.
-   Improve SimpleCov support.
  -   Very high test coverage.
-   Add Rubocop.

### CODE REVIEW

**Comment by halostatue** [comment]
This PR is currently WIP until the other gems are *also* updated.

File: lib/cartage/config.rb [hunk 1]
```diff
@@ -39,124 +73,72 @@ class Cartage
...
-          object.map! { |i| ostructify_recursively(i) }
-        when ::OpenStruct
-          object = ostructify(object.to_h)
-        when ::Hash
-          object = ostructify(object)
+          object.map { |e| ostructify(e) }
+        when ::OpenStruct, ::Hash
+          OpenStruct.new({}.tap { |h|
+            object.each_pair { |k, v| h[k] = ostructify(v) }
+          })
```
Comment (llwire) [inline lib/cartage/config.rb:138]: Nice

File: lib/cartage/gli_ext.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  # Mark the provided command as hidden from documentation.
+  def hide!
+    singleton_class.send(:define_method, :nodoc) { true }
+  end
+
+  def plugin_version_command(*plugin_classes)
+    desc 'Show the plug-in version'
+    command 'version' do |version|
+      version.hide!
+      version.action do |_g, _o, _a|
```
Comment (llwire) [inline lib/cartage/gli_ext.rb:63]: `_g, _o, _a` aren't referenced in the body

**Comment by llwire** [comment]
+1

Honestly, this is crazy good.

**Pull request merged**